sdk/client.py

```diff
@@ -147,20 +43,58 @@ def __init__(
         token_provider: Optional[TokenProvider] = None,
         endpoints: Optional[Endpoints] = None,
     ):
-        if not (username and password) and not token_provider:
-            raise Exception(
-                "At least a username/password combination or"
-                "TokenProvider object should be provided."
+        if not username and not token_provider:
+            raise ValueError(
+                "At least a username or TokenProvider object should be provided."
             )
-        self.endpoints = endpoints or Endpoints()
-        self.group_id = group_id
+        self._check_token_provider(token_provider)
 
-        if username and password:
-            token_provider = UsernamePasswordTokenProvider(
-                username, password, f"{self.endpoints.account}/api/v1/auth/login"
-            )
+        if username:
+            token_provider = self._credential_login(username, password)
 
         self.authenticator = HTTPBearerAuthenticator(token_provider)
+        self.endpoints = self._make_endpoints(endpoints)
+        self.group_id = group_id
+
+    @staticmethod
+    def _make_endpoints(endpoints: Optional[Endpoints]) -> Endpoints:
+        if endpoints is None:
+            return Endpoints()
+
+        if not isinstance(endpoints, Endpoints):
+            raise ValueError("Endpoints must be a Endpoints instance")
+
+        return endpoints
+
+    @staticmethod
+    def _check_token_provider(token_provider: Optional[TokenProvider]) -> None:
+        if not token_provider:
+            return
+
+        err = ValueError("token_provider must be a TokenProvider subclass")
+        try:
+            # The type ignore is because I wouldn't know how to fix the type problem
+            # but the code should be correct, and is tested
+            if not issubclass(token_provider, TokenProvider):  # type: ignore
+                raise err
+        except (
+            TypeError
+        ):  # This is the error if token_provider is not a class for issubclass
+            raise err
```

Comment:
Why not simply return the TypeError raised by issubclass ?

If we don't like that error message, and want our own `ValueError`, something like below sounds sufficient and simpler:
```python
try:
  issubclass(token_provider, TokenProvider)
except TypeError:
  raise ValueError(<error message>)
```

According to the [doc](https://docs.python.org/3/library/functions.html#issubclass), the `issubclass` function returns True or raises an error so the if statement looks unnecessary to me IIUC

Reply:
i dont see why i shouldn't do that, yeah :+1: i'll do that

Reply:
done